### PR TITLE
Port defines required for dynamic recompiler to Meson

### PR DESCRIPTION
**Pull request by dreamer**
I was testing platform builds - s390x and PPC64LE build fine, but ARM builds fail for some reason I do not understand yet… I don't know if it was autoconf doing something special or build.sh script.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -45,6 +45,63 @@ if get_option('enable_debugger') == 'heavy'
...
+  opts_for_arch = line[1]
+  dynrec_define = line[2]
+  target_cpu = line[3]
+
+  if (host_cpu == cpu) and opts_for_arch.contains(selected_core)
+    conf_data.set('C_TARGETCPU', target_cpu)
+    conf_data.set(dynrec_define, 1)
+    cpu_handles_unaligned_memory = true
+  endif
+endforeach
```
Comment (kcgen) [inline meson.build:91]: The foreach crawl the `core_selection` list looking for match. 
When a match is found (`host_cpu == cpu` and maching `arch`), should we break-out of the for-loop?

File: meson.build [hunk 2]
```diff
@@ -45,6 +45,63 @@ if get_option('enable_debugger') == 'heavy'
...
+
+conf_data.set10('C_UNALIGNED_MEMORY', cpu_handles_unaligned_memory)
```
Comment (kcgen) [inline meson.build:93]: Line 63 has: `cpu_handles_unaligned_memory = false`
So this statement here always sets `false` for handling `C_UNALIGNED_MEMORY`?

Or maybe the unaligned memory handling ability per architecture will be added to the `core_selection` table?

File: meson.build [hunk 3]
```diff
@@ -45,6 +45,63 @@ if get_option('enable_debugger') == 'heavy'
...
+
+target_cpu = conf_data.get('C_TARGETCPU')
```
Comment (kcgen) [inline meson.build:95]: If the above for-loop search for `host_cpu == cpu` wasn't found, then `conf_data.get('C_TARGETCPU')` will be empty.

Maybe this doesn't matter though? If conf_data.get('C_TARGETCPU') is empty and neither DYNAMIC* or DYNREC* are available, then we just get vanilla "normal-core only" build?

File: .github/workflows/linux.yml [hunk 4]
```diff
@@ -37,21 +37,24 @@ jobs:
...
+          - name: GCC, -dyn-x86
+            os: ubuntu-20.04
+            packages: meson
+            build_flags: -Ddynamic_core=dynrec
+            max_warnings: 0
+          - name: GCC, -dyn-x86, +debugger
+            os: ubuntu-20.04
+            packages: meson
+            build_flags: -Ddynamic_core=dynrec -Denable_debugger=normal
+            max_warnings: 144
```
Comment (kcgen) [inline .github/workflows/linux.yml]: This is really nice work @dreamer !

File: .github/workflows/platforms.yml [hunk 5]
```diff
@@ -52,14 +52,16 @@ jobs:
...
+          arch:   ${{ matrix.conf.arch }}
+          distro: ${{ matrix.conf.distro }}
           run: |
-            set -x
+            set -xo pipefail
             apt-get update
-            apt-get install -y libpng-dev $(./scripts/list-build-dependencies.sh -c gcc -m apt)
+            apt-get install -y $(cat ./.github/packages/ubuntu-18.04-apt.txt)
             ./scripts/log-env.sh
-            ./scripts/build.sh -c gcc -t release -m lto --disable-fluidsynth
```
Comment (kcgen) [inline .github/workflows/platforms.yml:62]: This is great - one step closer to having meson running the show!

**Review by kcgen — COMMENTED** [review]
Only some questions on the initial core selection logic. 

You're scaling meson to cover all of our CI needs nicely; not a small feat given we grew it organically over a much longer time-frame!

File: meson.build (hunk 1 above)
Comment (dreamer) [reply meson.build:91]: Initially, I had `break` in here, but the keyword was added in Meson 0.49.0, so it's unavailable in meson from Ubuntu 18.04 repo.

When I tried to disregard this problem and use meson installed via pip3, it resulted in problems for builds in `platforms` workflow :( It really is much easier to stick to tools available via official repositories.

For now, I decided to skip breaking out of the loop in here. I think I'll ask Meson maintainer (who is also packager for Ubuntu) to backport newer meson version to 18.04.

File: meson.build (hunk 2 above)
Comment (dreamer) [reply meson.build:93]: It sets `true` (which is converted to `1` in define due to usage of `set10` function).

All architectures we have defined ATM are marked as supporting unaligned memory in autoconf. We do build for s390x - I doubt anyone actually uses this arch for running DOSBox, but we are verifying codepaths for arch requiring aligned memory this way.

Old buildsystem marks these two options as *not* supporting unaligned memory:

```
   armv6l)
    AC_DEFINE(C_TARGETCPU,ARMV4LE)
    AC_MSG_RESULT(ARMv6 Little Endian)
    c_targetcpu="arm"
    dnl c_unalignedmemory=yes
    ;;
```
and
```
   *)
    AC_DEFINE(C_TARGETCPU,UNKNOWN)
    AC_MSG_RESULT(unknown)
    c_unalignedmemory=no
```

"armv6l" is weird and problematic ATM - the code was written for "ARMv4 Little Endian":

1) First problem is: meson does not differentiate between so old ARM versions - it reports 32-bit as `arm` (which is usually ARMv7) and 64-bit ARMv8 as `aarch64`. My guess is that we should probably move detection of older ARM variants to the code itself.
2) I do not understand the mess around current ARM code in cpu module - the code is written in several variants for "ARMv4", which is then supposed to be the same as ARMv6.

I couldn't find any references to ARMv4 - maybe it was some old version from 90s or something. ARMv6 is architecture from early 00s - still used in some embedded controllers and the like. The only device that we might want to support using this arch is Raspberry Pi 1 (but I do own RPi1 and tried to run dosbox on it once - it was starting up for 15 minutes and was totally unusable).

File: meson.build (hunk 3 above)
Comment (dreamer) [reply meson.build:95]: > Maybe this doesn't matter though? If conf_data.get('C_TARGETCPU') is empty and neither DYNAMIC* or DYNREC* are available, then we just get vanilla "normal-core only" build?

That's correct :)

When variable is not defined, then `#mesondefine C_TARGETCPU` in config.h.in file will evaluate to:

    /* #undef C_TARGETCPU */

And this define is used exclusively inside `src/cpu/`.

Majority of code in this PR is relevant to `src/cpu/` directory only… I am thinking about moving the logic in there to keep our top-level meson file a bit cleaner.

File: meson.build (hunk 2 above)
Comment (dreamer) [reply meson.build]: I moved code working with CPU-related configure defines to `src/cpu/meson.build` and included unaligned memory in the table - hopefully it's more readable now.

File: meson.build (hunk 3 above)
Comment (dreamer) [reply meson.build]: I moved the code and reorganized setting these defines in conf_data - should be more readable now.

File: meson.build (hunk 2 above)
Comment (kcgen) [reply meson.build]: Thanks for this explanation. It's fantastic CI does support these more obscure architectures.

The reorganized code looks great too.

**Review by kcgen — APPROVED** [review]
One step closer!

**Comment by dreamer** [comment]
For reference: #854

**Pull request merged**